File: Step 1 - Learn the Basics/1.3 Learn STL/01 Java Collections/SetExample.java
/**
 * Set Collection Example
 * 
 * Key Characteristics:
 * 1. Unique Elements: Does not allow duplicates
 * 2. At most one null element
 * 3. No guaranteed order (for HashSet)
 * 
 * Common Implementations:
 * - HashSet: Uses HashMap internally, no ordering, O(1) operations
 * - TreeSet: Sorted set, maintains natural order, O(log n) operations
 * - LinkedHashSet: Maintains insertion order, slightly slower than HashSet
 * 
 * Best Used When:
 * - Need to ensure element uniqueness
 * - Don't need to maintain element order (for HashSet)
 * - Need to quickly check if an element exists
 * - Need to eliminate duplicates from a collection
 */

import java.util.HashSet;
import java.util.Set;
import java.util.TreeSet;
import java.util.LinkedHashSet;
import java.util.Iterator;

public class SetExample {
    public static void main(String[] args) {
        // Example 1: HashSet
        System.out.println("=== HashSet Example ===");
        // Output: === HashSet Example ===
        
        Set<String> hashSet = new HashSet<>();
        
        // Adding elements
        hashSet.add("Apple");
        hashSet.add("Banana");
        hashSet.add("Apple");  // Duplicate element, will not be added
        hashSet.add("Orange");
        
        // Checking set state
        System.out.println("Set size: " + hashSet.size());
        // Output: Set size: 3
        
        System.out.println("Is set empty? " + hashSet.isEmpty());
        // Output: Is set empty? false
        
        System.out.println("Contains 'Banana'? " + hashSet.contains("Banana"));
        // Output: Contains 'Banana'? true
        
        // Iterating through set
        System.out.println("\nIterating through HashSet:");
        // Output: Iterating through HashSet:
        
        Iterator<String> iterator = hashSet.iterator();
        while (iterator.hasNext()) {
            System.out.println(iterator.next());
        }
        // Output: (order may vary)
        // Apple
        // Banana
        // Orange
        
        // Removing elements
        hashSet.remove("Banana");
        System.out.println("\nAfter removing 'Banana': " + hashSet);
        // Output: After removing 'Banana': [Apple, Orange]
        
        // Example 2: TreeSet (Sorted)
        System.out.println("\n=== TreeSet Example ===");
        // Output: === TreeSet Example ===
        
        Set<String> treeSet = new TreeSet<>();
        treeSet.add("Zebra");
        treeSet.add("Apple");
        treeSet.add("Banana");
        
        System.out.println("TreeSet (sorted): " + treeSet);
        // Output: TreeSet (sorted): [Apple, Banana, Zebra]
        
        // Example 3: LinkedHashSet (Maintains insertion order)
        System.out.println("\n=== LinkedHashSet Example ===");
        // Output: === LinkedHashSet Example ===
        
        Set<String> linkedHashSet = new LinkedHashSet<>();
        linkedHashSet.add("First");
        linkedHashSet.add("Second");
        linkedHashSet.add("Third");
        
        System.out.println("LinkedHashSet (maintains order): " + linkedHashSet);
        // Output: LinkedHashSet (maintains order): [First, Second, Third]
        
        // Set operations
        System.out.println("\n=== Set Operations ===");
        // Output: === Set Operations ===
        
        Set<String> set1 = new HashSet<>();
        set1.add("A");
        set1.add("B");
        set1.add("C");
        
        Set<String> set2 = new HashSet<>();
        set2.add("B");
        set2.add("C");
        set2.add("D");
        
        // Union
        Set<String> union = new HashSet<>(set1);
        union.addAll(set2);
        System.out.println("Union: " + union);
        // Output: Union: [A, B, C, D]
        
        // Intersection
        Set<String> intersection = new HashSet<>(set1);
        intersection.retainAll(set2);
        System.out.println("Intersection: " + intersection);
        // Output: Intersection: [B, C]
        
        // Difference
        Set<String> difference = new HashSet<>(set1);
        difference.removeAll(set2);
        System.out.println("Difference: " + difference);
        // Output: Difference: [A]
        
        // Clearing the set
        hashSet.clear();
        System.out.println("\nSet after clearing: " + hashSet);
        // Output: Set after clearing: []
    }
}

/**
 * Set Methods Summary Table:
 * 
 * | Category    | Method                    | Description                                    | Time Complexity | Throws Exception |
 * |-------------|---------------------------|------------------------------------------------|-----------------|------------------|
 * | Adding      | add(E e)                  | Adds element if not present                    | O(1) HashSet    | No               |
 * |             | addAll(Collection<?>)     | Adds all elements from collection              | O(n) HashSet    | Yes              |
 * | Removing    | remove(Object o)          | Removes specified element                      | O(1) HashSet    | No               |
 * |             | removeAll(Collection<?>)  | Removes all elements in collection             | O(n) HashSet    | Yes              |
 * |             | retainAll(Collection<?>)  | Keeps only elements in collection              | O(n) HashSet    | Yes              |
 * |             | clear()                   | Removes all elements                           | O(n)            | No               |
 * | Viewing     | contains(Object o)        | Checks if element exists                       | O(1) HashSet    | No               |
 * |             | containsAll(Collection<?>)| Checks if all elements exist                   | O(n) HashSet    | Yes              |
 * | Utility     | size()                    | Returns number of elements                     | O(1)            | No               |
 * |             | isEmpty()                 | Checks if set is empty                         | O(1)            | No               |
 * |             | toArray()                 | Converts set to array                          | O(n)            | No               |
 * 
 * Implementation-Specific Notes:
 * 
 * 1. HashSet
 *    - Fastest implementation (O(1) operations)
 *    - No ordering guarantee
 *    - Uses hashCode() and equals() for uniqueness
 *    - Best for general-purpose use
 * 
 * 2. TreeSet
 *    - Maintains natural ordering
 *    - Slower operations (O(log n))
 *    - Elements must be Comparable
 *    - Best for sorted data
 * 
 * 3. LinkedHashSet
 *    - Maintains insertion order
 *    - Slightly slower than HashSet
 *    - Uses more memory
 *    - Best when order matters
 * 
 * Common Use Cases:
 * 1. Duplicate Removal
 *    - Removing duplicates from collections
 *    - Unique value tracking
 *    - Distinct element counting
 * 
 * 2. Membership Testing
 *    - Quick lookups
 *    - Element existence checking
 *    - Unique identifier storage
 * 
 * 3. Set Operations
 *    - Union of collections
 *    - Intersection of collections
 *    - Difference between collections
 */

File: Step 1 - Learn the Basics/1.1 Things to Know/05 Arrays and Strings/SetExample.py
# Set Example with For Loop

my_set = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

print("Set elements squared:")
for item in my_set:
    print(item ** 2)

# Set Methods and Operations in Python

# Creating sets
set1 = {1, 2, 3, 4, 5}
set2 = {4, 5, 6, 7, 8}

print("Original sets:")
print("set1:", set1)  # Output: set1: {1, 2, 3, 4, 5}
print("set2:", set2)  # Output: set2: {4, 5, 6, 7, 8}

# 1. Adding elements
set1.add(6)
print("\nAfter adding 6 to set1:", set1)  # Output: After adding 6 to set1: {1, 2, 3, 4, 5, 6}

# 2. Removing elements
set1.remove(1)  # Raises KeyError if element not found
print("After removing 1 from set1:", set1)  # Output: After removing 1 from set1: {2, 3, 4, 5, 6}

set1.discard(10)  # No error if element not found
print("After trying to discard 10 from set1:", set1)  # Output: After trying to discard 10 from set1: {2, 3, 4, 5, 6}

# 3. Set Operations
# Union
union_set = set1 | set2  # or set1.union(set2)
print("\nUnion of set1 and set2:", union_set)  # Output: Union of set1 and set2: {2, 3, 4, 5, 6, 7, 8}

# Intersection
intersection_set = set1 & set2  # or set1.intersection(set2)
print("Intersection of set1 and set2:", intersection_set)  # Output: Intersection of set1 and set2: {4, 5, 6}

# Difference
difference_set = set1 - set2  # or set1.difference(set2)
print("Difference of set1 and set2:", difference_set)  # Output: Difference of set1 and set2: {2, 3}

# Symmetric Difference
symmetric_diff = set1 ^ set2  # or set1.symmetric_difference(set2)
print("Symmetric difference of set1 and set2:", symmetric_diff)  # Output: Symmetric difference of set1 and set2: {2, 3, 7, 8}

# 4. Set Methods
# Check if element exists
print("\nIs 3 in set1?", 3 in set1)  # Output: Is 3 in set1? True

# Check if set is subset
subset = {2, 3}
print("Is {2, 3} a subset of set1?", subset.issubset(set1))  # Output: Is {2, 3} a subset of set1? True

# Check if set is superset
print("Is set1 a superset of {2, 3}?", set1.issuperset(subset))  # Output: Is set1 a superset of {2, 3}? True

# Clear all elements
set3 = {1, 2, 3}
set3.clear()
print("\nAfter clearing set3:", set3)  # Output: After clearing set3: set()

# Copy a set
set4 = set1.copy()
print("Copy of set1:", set4)  # Output: Copy of set1: {2, 3, 4, 5, 6}

# Length of set
print("Length of set1:", len(set1))  # Output: Length of set1: 5

# Pop an arbitrary element
popped_element = set1.pop()
print("Popped element from set1:", popped_element)  # Output: Popped element from set1: 2 (or any other element)
print("set1 after popping:", set1)  # Output: set1 after popping: {3, 4, 5, 6} (remaining elements)

"""
Table of Set Methods Used in This Program:

| Method/Operation | Description                                    | Example                    | Return Value                    |
|-----------------|------------------------------------------------|----------------------------|--------------------------------|
| add()           | Adds a single element to the set              | set1.add(6)               | None                           |
| remove()        | Removes an element (raises KeyError if not found) | set1.remove(1)        | None                           |
| discard()       | Removes an element (no error if not found)    | set1.discard(10)          | None                           |
| union()         | Combines elements from both sets              | set1 | set2 or set1.union(set2) | New set with all elements     |
| intersection()  | Returns common elements                       | set1 & set2 or set1.intersection(set2) | New set with common elements |
| difference()    | Returns elements in first set but not in second | set1 - set2 or set1.difference(set2) | New set with unique elements |
| symmetric_difference() | Returns elements in either set but not both | set1 ^ set2 or set1.symmetric_difference(set2) | New set with unique elements |
| in              | Checks if element exists in set               | 3 in set1                 | Boolean (True/False)           |
| issubset()      | Checks if set is subset of another            | subset.issubset(set1)     | Boolean (True/False)           |
| issuperset()    | Checks if set is superset of another          | set1.issuperset(subset)   | Boolean (True/False)           |
| clear()         | Removes all elements from set                 | set3.clear()              | None                           |
| copy()          | Creates a shallow copy of set                 | set4 = set1.copy()        | New set with same elements     |
| len()           | Returns number of elements                    | len(set1)                 | Integer                        |
| pop()           | Removes and returns arbitrary element         | set1.pop()                | Removed element                |
"""
